### PR TITLE
improve identity handling

### PR DESCRIPTION
**Pull request by tomerd**
motivation: package identity is critical to SwiftPM ability to dedup package reliably

changes:
* use package identity instead of manifest name to flatten the dependencies list
* add more test for duplicate identity permutations across identity derived from URL and manifest name

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -74,19 +74,21 @@ extension PackageGraph {
...
                 )
-                flattenedManifests[node.manifest.name] = merged
+                flattenedManifests[packageIdentity] = merged
             } else {
-                flattenedManifests[node.manifest.name] = node
+                flattenedManifests[packageIdentity] = node
             }
         }
-        allManifests = flattenedManifests.values.sorted(by: { $0.manifest.name < $1.manifest.name })
+        allManifests = flattenedManifests.values.sorted(by: { identityResolver.resolveIdentity(for: $0.manifest.packageLocation) < identityResolver.resolveIdentity(for: $1.manifest.packageLocation) })
```
Comment (abertelrud) [inline Sources/PackageGraph/PackageGraph+Loading.swift:91]: How expensive is the identity resolution (in theory, once it's complete, regardless of the current implementation)?  Should it be pulled out of the sort condition here?

Comment (tomerd) [reply Sources/PackageGraph/PackageGraph+Loading.swift:91]: good point, eventually it will be cached by the resolver itself.

**Review by neonichu — APPROVED** [review]
